### PR TITLE
[MIG]12.0: Migrated Jasper Report to v12.

### PR DESCRIPTION
**Pull request by hiren-serpentcs**

### CODE REVIEW

**Comment by JayVora-SerpentCS** [comment]
@swapnesh-serpentcs do a tech review

File: jasper_reports/JasperReports/jasper_report.py [hunk 1]
```diff
@@ -280,8 +279,8 @@ def extract_properties(self):
 
             # Relations
             relations = []
-            path8 = '../../jr:reportElement/jr:property \
-            [@name="OPENERP_RELATIONS"]'
+            path8 = \
+                '../../jr:reportElement/jr:property[@name="OPENERP_RELATIONS"]'
```
Comment (JayVora-SerpentCS) [inline jasper_reports/JasperReports/jasper_report.py]: we must rename all key params as odoo-oriented not openerp

File: jasper_load/__manifest__.py [hunk 2]
```diff
@@ -29,11 +29,12 @@
```
Comment (JayVora-SerpentCS) [inline jasper_load/__manifest__.py:29]: index.html pan update karjo.. mention specially k jasper_load is necessary... u know ne.. public shout

Comment (Swapnesh-SerpentCS) [reply jasper_load/__manifest__.py:29]: and also All those Points which we had to reply on Github Issue/Odoo App page.

File: jasper_reports/JasperReports/browse_data_generator.py [hunk 3]
```diff
@@ -35,14 +35,13 @@
 
 import os
 import csv
-# import copy
 import base64
 from xml.dom.minidom import getDOMImplementation
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/JasperReports/browse_data_generator.py]: Move this below

File: jasper_reports/JasperReports/browse_data_generator.py [hunk 4]
```diff
@@ -155,21 +146,21 @@ def generate_ids(self, record, relations, path, current_records):
...
                 current_records = new_records
         return current_records
 
 
 class XmlBrowseDataGenerator(BrowseDataGenerator):
     def __init__(self, report, model, env, cr, uid, ids, context):
-        super(XmlBrowseDataGenerator, self).__init__(report, model, env, cr,
-                                                     uid, ids, context)
+        super(XmlBrowseDataGenerator, self).__init__(
+            report, model, env, cr, uid, ids, context)
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/JasperReports/browse_data_generator.py:163]: not sure if this is valid but Do we have this according to new API version?

File: jasper_reports/JasperReports/jasper_report_config.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+# Copyright (c) 2008-2012 NaN Projectes de Programari Lliure, S.L.
+#                         http://www.NaN-tic.com
+# Copyright (C) 2013 Tadeus Prastowo <tadeus.prastowo@infi-nity.com>
+#                         Vikasa Infinity Anugrah <http://www.infi-nity.com>
+# Copyright (C) 2011-Today Serpent Consulting Services Pvt. Ltd.
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/JasperReports/jasper_report_config.py]: 2019

File: jasper_reports/JasperReports/jasper_report_config.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            report_module = path.split(os.path.sep)[0]
+
+            for addons_path in tools.config['addons_path'].split(','):
+                if os.path.lexists(addons_path + os.path.sep + report_module):
+                    return os.path.normpath(addons_path + os.path.sep + path)
+
+        return os.path.dirname(self.path())
+
+    def system_user_name(self):
+        if os.name == 'nt':
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/JasperReports/jasper_report_config.py:216]: Maybe better to use `sys_platform`

File: jasper_reports/JasperReports/jasper_report_config.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+    def user_name(self):
+        # To allow all users to get db_user we have to call it with sudo
+        user_name = self.env['ir.config_parameter'].sudo().get_param(
+            'db_user') or self.system_user_name()
+        return tools.config['db_user'] or user_name
+
+    def password(self):
+        # To allow all users to get db_password we have to call it with sudo
+        password = self.env['ir.config_parameter'].sudo().get_param(
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/JasperReports/jasper_report_config.py:237]: `db_user` and `db_password` can be fetched from odoo config file, Check this in odoo/tools.config .py

File: jasper_reports/JasperReports/jasper_server.py [hunk 8]
```diff
@@ -72,7 +72,7 @@ def start(self):
                 self.javapath = java_path
             else:
                 raise UserError(_('libraries Not Found !\n'
-                                'There is No libraries found in Java'))
+                                  'There is No libraries found in Java'))
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/JasperReports/jasper_server.py]:
```suggestion
                                  'No libraries found in Java.'))
```

File: jasper_reports/__manifest__.py [hunk 9]
```diff
@@ -49,17 +46,16 @@
...
-        'views/webclient_templates.xml',
-        'jasper_report_menu.xml',
-        'wizard/jasper_create_data_template.xml',
-        'jasper_wizard.xml',
-        'report_xml_view.xml',
         'security/ir.model.access.csv',
         'data/jasper_data.xml',
-        'res_company_view.xml',
+        'views/webclient_templates.xml',
+        'wizard/jasper_create_data_template.xml',
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/__manifest__.py:54]: wizard should be first (If not needed strictly below )

File: jasper_reports/data/jasper_data.xml [hunk 10]
```diff
@@ -1,13 +1,11 @@
...
-            <field name="key">db_user</field>
-            <field name="value">serpentcs</field>
-        </record>
-        <record id="db_password_ir_config_parameter" model="ir.config_parameter">
-            <field name="key">db_password</field>
-            <field name="value">serpentcs</field>
-        </record>
-    </data>
-</openerp>
+<odoo>
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/data/jasper_data.xml]: noupdate=1

File: jasper_reports/models/__init__.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+# Copyright (c) 2008-2012 NaN Projectes de Programari Lliure, S.L.
+#                         http://www.NaN-tic.com
+# Copyright (C) 2013 Tadeus Prastowo <tadeus.prastowo@infi-nity.com>
+#                         Vikasa Infinity Anugrah <http://www.infi-nity.com>
+# Copyright (C) 2011-Today Serpent Consulting Services Pvt. Ltd.
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/__init__.py]: 2019

File: jasper_reports/models/report_xml.py [hunk 12]
```diff
@@ -90,39 +85,39 @@ def write(self, values):
...
-    jasper_file_ids = fields.One2many('ir.actions.report.xml.file',
-                                      'report_id', 'Files', help='')
+    jasper_output = fields.Selection(
+        [('html', 'HTML'), ('csv', 'CSV'),
+         ('xls', 'XLS'), ('rtf', 'RTF'),
+         ('odt', 'ODT'), ('ods', 'ODS'),
+         ('txt', 'Text'), ('pdf', 'PDF')],
+        default='pdf')
+    jasper_file_ids = fields.One2many(
+        'ir.actions.report.xml.file', 'report_id', 'Files', help='')
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]:
```suggestion
        'ir.actions.report.xml.file', 'report_id', 'Files')
```
or add some help

File: jasper_reports/models/report_xml.py [hunk 13]
```diff
@@ -90,39 +85,39 @@ def write(self, values):
...
-        ], limit=1)
-
-    @api.multi
+        attachment_obj = self.env['ir.attachment']
+        for report in self:
+            if report.attachment:
+                attachment_name = safe_eval(
+                    report.attachment, {'object': record, 'time': time})
+            else:
+                attachment_name = str(report.name) + '.' + report.jasper_output
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]: Remove else and take `attachment_name = str(report.name) + '.' + report.jasper_output` above

File: jasper_reports/models/report_xml.py [hunk 14]
```diff
@@ -133,40 +128,44 @@ def postprocess_jasper_report(self, record, buffer):
...
-        else:
-            _logger.info('The %s document %s is now saved in the database',
-                         self.jasper_output, attachment_vals['name'])
-        return attachment
+        attachment_obj = self.env['ir.attachment']
+        for report in self:
+            if report.attachment:
+                attachment_name = safe_eval(
+                    report.attachment, {'object': record, 'time': time})
+            else:
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]: same as above

File: jasper_reports/models/report_xml.py [hunk 15]
```diff
@@ -133,40 +128,44 @@ def postprocess_jasper_report(self, record, buffer):
...
+            attachment_vals = {
+                'name': attachment_name,
+                'datas': base64.encodestring(buffer.getvalue()),
+                'datas_fname': attachment_name,
+                'res_model': report.model,
+                'res_id': record.id,
+            }
+            attachment = None
+            try:
+                attachment = attachment_obj.create(attachment_vals)
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]:
```suggestion
                return attachment_obj.create(attachment_vals)
```

File: jasper_reports/models/report_xml.py [hunk 16]
```diff
@@ -133,40 +128,44 @@ def postprocess_jasper_report(self, record, buffer):
...
+            except AccessError:
+                _logger.info(
+                    "Cannot save %s report %r as attachment",
+                    report.jasper_output, attachment_vals['name'])
+            else:
+                _logger.info('The %s document %s is now saved in the database',
+                             report.jasper_output, attachment_vals['name'])
+            return attachment
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]:
```suggestion
            return None
```

File: jasper_reports/models/report_xml.py [hunk 17]
```diff
@@ -133,40 +128,44 @@ def postprocess_jasper_report(self, record, buffer):
...
+                'datas': base64.encodestring(buffer.getvalue()),
+                'datas_fname': attachment_name,
+                'res_model': report.model,
+                'res_id': record.id,
+            }
+            attachment = None
+            try:
+                attachment = attachment_obj.create(attachment_vals)
+            except AccessError:
+                _logger.info(
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]:
```suggestion
                _logger.warn(
```

File: jasper_reports/models/report_xml.py [hunk 18]
```diff
@@ -175,24 +174,23 @@ def render_jasper(self, docids, data):
         record = self.env.cr.dictfetchone()
         report_model = self.search([('report_name', '=', report_model_name)])
         if report_model is None:
-            raise UserError(_('%s model was not found' % report_model_name))
+            raise UserError(_('%s model was not found') % report_model_name)
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]:
```suggestion
            raise UserError(_('%s model not found.') % report_model_name)
```

File: jasper_reports/models/report_xml.py [hunk 19]
```diff
@@ -175,24 +174,23 @@ def render_jasper(self, docids, data):
...
 
     @api.model
     def _get_report_from_name(self, report_name):
         res = super(ReportXml, self)._get_report_from_name(report_name)
         if res:
             return res
         report_obj = self.env['ir.actions.report']
-        qwebtypes = ['jasper']
-        conditions = [('report_type', 'in', qwebtypes),
+        conditions = [('report_type', '=', 'jasper'),
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py]: domain

File: jasper_reports/models/report_xml.py [hunk 20]
```diff
@@ -201,26 +199,24 @@ def _get_report_from_name(self, report_name):
...
-                        values['jasper_model_id']).model
+                        values['model_id']).model
 
             values['type'] = 'ir.actions.report'
             values['report_type'] = 'jasper'
             values['jasper_report'] = True
         return super(ReportXml, self).write(values)
 
-    @api.multi
     def update(self):
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/models/report_xml.py:217]: Remove this and do directly using api.constraint

Two methods for each field

File: jasper_reports/views/jasper_report_menu.xml [hunk 21]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0"?>
+<odoo>
+    <menuitem parent="base.menu_custom" name="Jasper Reports" id="jasper_reports_menu"/>
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/views/jasper_report_menu.xml]: Indentation

File: jasper_reports/views/report_xml_view.xml [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+
+    <record id="act_report_jasper_form" model="ir.ui.view">
+        <field name="name">ir.actions.report.jasper.form</field>
+        <field name="model">ir.actions.report</field>
+        <field name="priority">20</field>
+        <field name="arch" type="xml">
+            <form string="Jasper Reports">
+                <group colspan="4" col='4'>
+                    <field name="name"/>
+                    <field name="model_id" required="True" readonly="False"/>
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/views/report_xml_view.xml:47]: Check readonly="False"

File: jasper_reports/wizard/jasper_create_data_template.py [hunk 23]
```diff
@@ -32,38 +32,34 @@
...
-            'view_mode': 'form',
-            'view_type': 'form',
-            'res_id': rec.id,
-            'views': [(False, 'form')],
-            'target': 'new',
-        }
-
-    model = fields.Many2one('ir.model', 'Model', required=True)
-    depth = fields.Integer("Depth", required=True, default=1)
+    model = fields.Many2one('ir.model', required=True)
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/wizard/jasper_create_data_template.py]:
```suggestion
    model_id = fields.Many2one('ir.model', required=True)
```

File: jasper_reports/wizard/jasper_create_data_template.py [hunk 24]
```diff
@@ -32,38 +32,34 @@
...
+    depth = fields.Integer(required=True, default=1)
     filename = fields.Char('File Name', size=32)
     data = fields.Binary('XML')
+
+    def action_create_xml(self):
+        for rec in self:
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/wizard/jasper_create_data_template.py]:
```suggestion
        for data_template in self:
```

File: jasper_reports/wizard/jasper_create_data_template.py [hunk 25]
```diff
@@ -32,38 +32,34 @@
...
+            xml = self.env['ir.actions.report'].create_xml(
+                rec.model.model, rec.depth)
+            base64_str = base64.encodestring(
+                ('%s' % (xml)).encode()).decode().replace('\n', '')
+            rec.write({
+                'data': base64_str,
+                'filename': str(rec.model.name) + '_template.xml'})
+            return {
```
Comment (Swapnesh-SerpentCS) [inline jasper_reports/wizard/jasper_create_data_template.py]: Instead of return, Read action and return by changing required keys

**Review by Swapnesh-SerpentCS — CHANGES_REQUESTED** [review]

**Review by Swapnesh-SerpentCS — APPROVED** [review]

**Pull request merged**